frontends/common/resolveReferences/resolveReferences.cpp

```diff
@@ -107,6 +107,18 @@ ResolutionContext::lookup(const IR::INamespace *current, IR::ID name,
                 Util::SourceInfo dsi = decl->getNode()->srcInfo;
                 bool before = dsi <= nsi;
                 LOG3("\tPosition test:" << dsi << "<=" << nsi << "=" << before);
+
+                // Type foo = foo (don't resolve rhs to lhs)
+                if (auto *decl_var = decl->to<IR::Declaration_Variable>()) {
```

Comment:
This  fix is very specific to variable declarations, but I suspect that the same problem exists for other declarations and instantiations.

Reply:
I've expanded the logic in _checkShadowing()_ to cover other parameter shadowing scenarios. As for the code in the _lookup()_ method, I must elaborate:

Since the bug seems to be occurring (at least for now) only when there is an identical name on both sides of the '=' sign, then our possible node candidates become:

1. Declaration_Variable (has _initializer_ as member)
2. Declaration_Constant (also has _initializer_ as member)
3. and you mentioned instantiation (which is used for Declaration_Instance).

The first 2 are now accounted for. As for 3. I have to ask - could you elaborate on how instantiation is involved ? (test examples are much appreciated) 
From what I've seen in the specification, instantiations don't have a right-hand side. Unless you're referring to the experimental feature (https://github.com/p4lang/p4c/blob/c0907946eeb46a96bc8fee822ba083fb3c442bd2/frontends/parsers/p4/p4parser.ypp#L673)

Reply:
Yes, we have to support the experimental feature too.
I haven't checked, but I expect that this could also apply to other declarations that introduce scopes, e.g., `struct S { S x; }`

Reply:
> I haven't checked, but I expect that this could also apply to other declarations that introduce scopes, e.g., `struct S { S x; }`

You were right, it does. I'm working on a fix at the moment.
> Yes, we have to support the experimental feature too.

I haven't had much luck in finding some examples that demonstrate use of this feature. Any short examples you could share ?